### PR TITLE
core: make default matrix follow the lite API

### PR DESCRIPTION
**Pull request by dexter93**
stock matrix_scan function is a glorified custom function by default. Time to actually be one
Make init and scan functions to follow the `lite` API. Allow their overriding by user-side code

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/matrix_common.c [hunk 1]
```diff
@@ -133,11 +133,8 @@ __attribute__((weak)) void matrix_output_unselect_delay(uint8_t line, bool key_p
     matrix_io_delay();
 }
 
-// CUSTOM MATRIX 'LITE'
-__attribute__((weak)) void matrix_init_custom(void) {}
-__attribute__((weak)) bool matrix_scan_custom(matrix_row_t current_matrix[]) {
-    return true;
-}
+__attribute__((weak)) void matrix_init_custom(void);
+__attribute__((weak)) bool matrix_scan_custom(matrix_row_t current_matrix[]);
```
Comment (zvecr) [inline quantum/matrix_common.c:137]:
```suggestion
void matrix_init_custom(void);
bool matrix_scan_custom(matrix_row_t current_matrix[]);
```

Comment (dexter93) [reply quantum/matrix_common.c:137]: Was thinking about this. Makes sense since it's the prototypes. Will test and report back

**Comment by tzarc** [comment]
Deferring to next quarter, mitigating risks with such a core part of the code, and no tangible benefit to be seen by getting it into `master` this cycle -- code that leverages it will end up being put through `develop` anyway.

**Comment by dexter93** [comment]
> Deferring to next quarter, mitigating risks with such a core part of the code, and no tangible benefit to be seen by getting it into `master` this cycle -- code that leverages it will end up being put through `develop` anyway.

does that mean it will get merged in `develop` soon, or it it deferred to the next breaking cycle into next `master` update?

**Comment by tzarc** [comment]
The `breaking_change_2023q1` label means it _should_ be included in the 2023q1 merge. If the stars align, it'll land in `develop` well before then.

**Comment by tzarc** [comment]
Internally discussed; we're going to be going with a `MATRIX_DRIVER = xxxx` solution in a general sense, will supersede this.

**Pull request closed without merging**